src/mbgl/style/expression/number_format.cpp

```diff
@@ -0,0 +1,291 @@
+#include <mbgl/style/conversion_impl.hpp>
+
+#include <locale>
+#include <sstream>
+#include <iomanip>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+NumberFormat::NumberFormat(std::unique_ptr<Expression> number_,
+                           optional<std::unique_ptr<Expression>> locale_,
+                           optional<std::unique_ptr<Expression>> currency_,
+                           optional<std::unique_ptr<Expression>> minFractionDigits_,
+                           optional<std::unique_ptr<Expression>> maxFractionDigits_)
+    : Expression(Kind::NumberFormat, type::String),
+      number(std::move(number_))
+{
+      if (locale_) {
+            locale = std::unique_ptr<Expression>(std::move(*locale_));
+      }
+      if (currency_) {
+            currency = std::unique_ptr<Expression>(std::move(*currency_));
+      }
+      if (minFractionDigits_) {
+            minFractionDigits = std::unique_ptr<Expression>(std::move(*minFractionDigits_));
+      }
+      if (maxFractionDigits_) {
+            maxFractionDigits = std::unique_ptr<Expression>(std::move(*maxFractionDigits_));
+      }
+}
+
+EvaluationResult NumberFormat::evaluate(const EvaluationContext& params) const {
+    auto numberResult = number->evaluate(params);
+    if (!numberResult) {
+        return numberResult.error();
+    }
+    double evaluatedNumber = numberResult->get<double>();
+    
+    optional<std::string> evaluatedLocale;
+    if (locale) {
+        auto localeResult = (*locale)->evaluate(params);
+        if (!localeResult) {
+            return localeResult.error();
+        }
+        evaluatedLocale = toString(*localeResult);
+        // Replace '-' with '_' in the locale string and append '.utf8' to it
+        (*evaluatedLocale).replace((*evaluatedLocale).find_first_of("-"), 1, "_");
+        (*evaluatedLocale).append(".utf8");
+    }
+
+    optional<std::string> evaluatedCurrency;
+    if (currency) {
+        auto currencyResult = (*currency)->evaluate(params);
+        if (!currencyResult) {
+            return currencyResult.error();
+        }
+        evaluatedCurrency = toString(*currencyResult);
+    }
+
+    optional<double> evaluatedMinFractionDigits;
+    if (minFractionDigits) {
+        auto minDigitsResult = (*minFractionDigits)->evaluate(params);
+        if (!minDigitsResult) {
+            return minDigitsResult.error();
+        }
+        evaluatedMinFractionDigits = minDigitsResult->get<double>();
+    }
+
+    optional<double> evaluatedMaxFractionDigits;
+    if (maxFractionDigits) {
+        auto maxDigitsResult = (*maxFractionDigits)->evaluate(params);
+        if (!maxDigitsResult) {
+            return maxDigitsResult.error();
+        }
+        evaluatedMaxFractionDigits = maxDigitsResult->get<double>();
+    }
+
+    std::stringstream ss;
+    std::locale loc;
+    bool localeSet;
+    try {
+        loc = std::locale(evaluatedLocale ? (*evaluatedLocale).c_str() : "");
```

Comment:
if `evaluatedLocale`were `std::string`, we'd had just `loc = std::locale(evaluatedLocale)`